kube/src/api/portforward.rs

```diff
@@ -0,0 +1,225 @@
+        }
+
+        let state = Arc::new(Mutex::new(PortforwarderState {
+            waker: None,
+            finished: false,
+            ports: Some(ports),
+        }));
+        let shared_state = state.clone();
+        let port_nums = port_nums.to_owned();
+        tokio::spawn(async move {
+            start_message_loop(stream, &port_nums, duplexes, errors).await;
+
+            let mut shared = shared_state.lock().unwrap();
+            shared.finished = true;
+            if let Some(waker) = shared.waker.take() {
+                waker.wake()
+            }
+        });
+        Portforwarder { state }
+    }
+
+    /// Get streams for forwarded ports.
+    pub fn ports(&mut self) -> Option<Vec<Port>> {
+        let mut state = self.state.lock().unwrap();
+        state.ports.take()
+    }
+}
+
+impl Future for Portforwarder {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut state = self.state.lock().unwrap();
+        if state.finished {
+            return Poll::Ready(());
+        }
+
+        if let Some(waker) = &state.waker {
+            if waker.will_wake(cx.waker()) {
+                return Poll::Pending;
+            }
+        }
+
+        state.waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+pub struct Port {
+    // Data pipe.
+    stream: Option<DuplexStream>,
+    // Error channel.
+    error: Option<ErrorReceiver>,
+}
+
+impl Port {
+    pub(crate) fn new(stream: DuplexStream, error: ErrorReceiver) -> Self {
+        Port {
+            stream: Some(stream),
+            error: Some(error),
+        }
+    }
+
+    /// Data pipe for sending to and receiving from the forwarded port.
+    pub fn stream(&mut self) -> Option<impl AsyncRead + AsyncWrite + Unpin> {
+        self.stream.take()
+    }
+
+    /// Future that resolves with any error message or when the error sender is dropped.
+    /// When this resolves, the port should be considered no longer usable.
+    pub fn error(&mut self) -> Option<impl Future<Output = Option<String>>> {
+        // Ignore Cancellation error.
+        self.error.take().map(|recv| recv.map(|res| res.ok()))
+    }
+}
+
+async fn start_message_loop<S>(
+    stream: WebSocketStream<S>,
+    ports: &[u16],
+    duplexes: Vec<DuplexStream>,
+    mut errors: Vec<Option<ErrorSender>>,
+) where
+    S: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+{
+    let mut writers = Vec::new();
+    let (tx, mut rx) = mpsc::channel::<Message>(1);
+    for (i, (r, w)) in duplexes.into_iter().map(tokio::io::split).enumerate() {
+        writers.push(w);
+        {
+            // Each port uses 2 channels. Duplex data channel and error.
+            let ch = 2 * (i as u8);
+            let mut read_stream = tokio_util::io::ReaderStream::new(r);
+            let mut txr = tx.clone();
+            tokio::spawn(async move {
+                while let Some(res) = read_stream.next().await {
```

Comment:
interface almost feels like a generalised attach, except we have a task per port.